### PR TITLE
[INT-109] Create dictionary entity

### PR DESCRIPTION
**Pull request by polina-koval**
https://huntflow.atlassian.net/browse/INT-109

### CODE REVIEW

File: huntflow_api_client/models/common.py [hunk 1]
```diff
@@ -5,5 +5,5 @@
 
 
 class JsonRequestModel(BaseModel):
-    def jsonable_dict(self) -> Dict[str, Any]:
-        return json.loads(self.json())
+    def jsonable_dict(self, exclude_none: bool = False) -> Dict[str, Any]:
```
Comment (alexreznikoff) [inline huntflow_api_client/models/common.py:8]: It seems you need to copy all the parameters of the `BaseModel.json()` method and pass them to `json` method
```python
class JsonRequestModel(BaseModel):
    def jsonable_dict(
        self,
        *,
        include: Union['AbstractSetIntStr', 'MappingIntStrAny'] = None,
        exclude: Union['AbstractSetIntStr', 'MappingIntStrAny'] = None,
        by_alias: bool = False,
        skip_defaults: bool = None,
        exclude_unset: bool = False,
        exclude_defaults: bool = False,
        exclude_none: bool = False,
        encoder: Optional[Callable[[Any], Any]] = None,
        **dumps_kwargs: Any,
    ) -> Dict[str, Any]:
        params = {
            "include": include,
            "exclude": exclude,
            "by_alias": by_alias,
            "skip_defaults": skip_defaults,
            "exclude_unset": exclude_unset,
            "exclude_defaults": exclude_defaults,
            "exclude_none": exclude_none,
            "encoder": encoder,
        }
        return json.loads(self.json(**params, **dumps_kwargs))
```

File: huntflow_api_client/models/utils/fields.py [hunk 2]
```diff
@@ -0,0 +1,20 @@
+from datetime import datetime
+from typing import Generator, Union
+
+from pydantic import datetime_parse
+from pydantic.datetime_parse import StrBytesIntFloat
+
+
+class DatetimeWithTZ(datetime):
```
Comment (alexreznikoff) [inline huntflow_api_client/models/utils/fields.py:8]: I think this class is useless, it is enough to use the usual `datetime` type

File: huntflow_api_client/api.py [hunk 3]
```diff
@@ -111,6 +111,7 @@ async def _request(  # type: ignore[no-untyped-def]
                 timeout=timeout,
             )
             await self._raise_token_expired(response)
+            response.raise_for_status()
```
Comment (alexreznikoff) [inline huntflow_api_client/api.py:114]: It's a bad idea to use default `response.raise_for_status()`

File: huntflow_api_client/entities/dictionaries.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
+from huntflow_api_client.entities.base import (
+    BaseEntity,
+    CreateEntityMixin,
+    GetEntityMixin,
+    ListEntityMixin,
+    UpdateEntityMixin,
```
Comment (alexreznikoff) [inline huntflow_api_client/entities/dictionaries.py:6]: all these mixins can be replaced with one `CRUDEntityMixin`

File: huntflow_api_client/entities/dictionaries.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+)
+from huntflow_api_client.models.response.dictionaries import (
+    DictionariesListResponse,
+    DictionaryCreateResponse,
+    DictionaryResponse,
+    DictionaryUpdateResponse,
+)
+
+
+class Dictionary(BaseEntity, UpdateEntityMixin, ListEntityMixin, CreateEntityMixin, GetEntityMixin):
```
Comment (alexreznikoff) [inline huntflow_api_client/entities/dictionaries.py:20]: `class Dictionary(BaseEntity, CRUDEntityMixin)`

File: huntflow_api_client/models/utils/descriptions.py [hunk 6]
```diff
@@ -1,3 +1,9 @@
 foreign = "The unique identifier in the customer's internal system"
 organization_id = "Organization ID"
 task_id = "Task ID"
+dict_id = "Dictionary ID"
```
Comment (alexreznikoff) [inline huntflow_api_client/models/utils/descriptions.py:4]: not sure if this file will be user-friendly at all

File: huntflow_api_client/models/response/dictionaries.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+from uuid import UUID
+
+from pydantic import BaseModel, Field, PositiveInt
+
+from huntflow_api_client.models.utils import descriptions
+from huntflow_api_client.models.utils.fields import DatetimeWithTZ
+
+
+class DictionaryTaskResponsePayload(BaseModel):
+    task_id: UUID = Field(..., description="Job ID")
```
Comment (alexreznikoff) [inline huntflow_api_client/models/response/dictionaries.py:11]: `task_id: UUID = Field(..., description="Task ID")`

File: huntflow_api_client/models/response/dictionaries.py [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+    status: str = Field(..., description="Operation status", example="ok")
+    payload: DictionaryTaskResponsePayload
+    meta: DictionaryTaskResponseMeta
+
+
+class DictionaryUpdateResponse(DictionaryTaskResponse):
+    pass
+
+
+class DictionaryCreateResponse(DictionaryTaskResponse):
```
Comment (alexreznikoff) [inline huntflow_api_client/models/response/dictionaries.py:29]: Do we really need two identical classes? `DictionaryUpdateResponse` and `DictionaryCreateResponse`
Why not use one `DictionaryTaskResponse` class in both cases

File: huntflow_api_client/models/request/dictionaries.py [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+    )
+    meta: Optional[dict] = Field(
+        None,
+        description=descriptions.meta,
+        example={"latitude": 55.5374, "longitude": 60.1408},
+    )
+    items: Optional[List[DictionaryItem]] = Field(None, description=descriptions.dict_items)
+
+
+DictionaryItem.update_forward_refs()
```
Comment (alexreznikoff) [inline huntflow_api_client/models/request/dictionaries.py:27]: useless
In API it was used to build swagger documentation

File: huntflow_api_client/models/response/dictionaries.py [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+    pass
+
+
+class DictionaryItem(BaseModel):
+    id: PositiveInt = Field(..., description=descriptions.dict_id, example=7)  # noqa: A003 VNE003
+    code: str = Field(..., description=descriptions.dict_code, example="citizenship")
+    name: str = Field(..., description=descriptions.dict_name, example="Citizenship")
+    foreign: Optional[str] = Field(None, description=descriptions.foreign, example="d_ctz")
+    created: DatetimeWithTZ = Field(..., description="Date and time of creating a dictionary")
```
Comment (alexreznikoff) [inline huntflow_api_client/models/response/dictionaries.py:38]: created: **datetime** = Field(..., description="Date and time of creating a dictionary")

File: huntflow_api_client/models/response/dictionaries.py [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+class DictionaryResponse(BaseModel):
+    id: int = Field(..., description=descriptions.dict_id)  # noqa: A003 VNE003
+    code: str = Field(..., description=descriptions.dict_code)
+    name: str = Field(..., description=descriptions.dict_name)
+    foreign: Optional[str] = Field(None, description=descriptions.foreign)
+    created: DatetimeWithTZ = Field(..., description="Date and time of creating a dictionary")
+    dictionary_fields: List[DictionaryField] = Field(
+        ...,
+        alias="fields",
+        description="Dictionary fields",
```
Comment (alexreznikoff) [inline huntflow_api_client/models/response/dictionaries.py:69]: description="List of dictionary fields",

File: huntflow_api_client/models/response/dictionaries.py [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+    meta: Optional[dict] = Field(
+        None,
+        description=descriptions.meta,
+        example={"latitude": 55.5374, "longitude": 60.1408},
+    )
+
+
+class DictionaryResponse(BaseModel):
+    id: int = Field(..., description=descriptions.dict_id)  # noqa: A003 VNE003
+    code: str = Field(..., description=descriptions.dict_code)
```
Comment (alexreznikoff) [inline huntflow_api_client/models/response/dictionaries.py:62]: Its not user-friendly descriptions: `descriptions.dict_code`, `descriptions.dict_id` and so on

File: huntflow_api_client/models/response/dictionaries.py [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+    name: str = Field(..., description=descriptions.dict_name)
+    foreign: Optional[str] = Field(None, description=descriptions.foreign)
+    created: DatetimeWithTZ = Field(..., description="Date and time of creating a dictionary")
```
Comment (alexreznikoff) [inline huntflow_api_client/models/response/dictionaries.py:65]: created: **datetime** = Field(..., description="Date and time of creating a dictionary")

File: huntflow_api_client/entities/dictionaries.py (hunk 4 above)
Comment (polina-koval) [reply huntflow_api_client/entities/dictionaries.py:6]: Actually, no. CRUDEntityMixin has DeleteEntityMixin. Instead, we need an UpdateEntityMixin.

File: huntflow_api_client/entities/dictionaries.py (hunk 5 above)
Comment (polina-koval) [reply huntflow_api_client/entities/dictionaries.py:20]: CRUDEntityMixin has DeleteEntityMixin. Instead, we need an UpdateEntityMixin.

File: huntflow_api_client/api.py (hunk 3 above)
Comment (polina-koval) [reply huntflow_api_client/api.py:114]: Why? What are the alternatives?

File: huntflow_api_client/models/request/dictionaries.py [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+class DictionaryCreateRequest(JsonRequestModel):
+    code: str = Field(..., description="Dictionary code", min_length=1, max_length=128)
+    name: str = Field(..., description="Dictionary name", min_length=1, max_length=255)
+    foreign: Optional[str] = Field(
+        None,
+        description="The unique identifier in the customer's internal system",
+        min_length=1,
+        max_length=1024,
+    )
+    items: List[DictionaryItem] = Field(..., description="Dictionary items")
```
Comment (alexreznikoff) [inline huntflow_api_client/models/request/dictionaries.py:33]: Maybe it's better to use a common name for this field?
In requests it is called `items`
In responses it is called `dictionary_fields`

**Pull request merged**